Assets/Scripts/SS3D/Systems/Substances/SubstanceContainerActor.cs

```diff
@@ -0,0 +1,618 @@
+        {
+            OnContentsChanged();
+        }
+
+        [Server]
+        protected virtual void OnContentsChanged()
+        {
+             ProcessContainer(this);
+            ContentsChanged?.Invoke(this);
+        }
+
+        public override IInteraction[] CreateTargetInteractions(InteractionEvent interactionEvent)
+        {
+            return new IInteraction[]
+            {
+                new TransferSubstanceInteraction()
+            };
+        }
+
+        /// <summary>
+        /// Processes substances in a container
+        /// </summary>
+        public static void ProcessContainer(SubstanceContainerActor container)
+        {
+            SubstanceContainer.ProcessContainer(container.substanceContainer);
+            container.MarkDirty();
+        }
+
+        #endregion
+
+        #region SubstanceContainer
+        public class SubstanceContainer
+        {
+            /// <summary>
+            /// do not modify directly, use the property setter Volume instead.
+            /// </summary>
+            private float _volume;
+
+            private float _currentVolume;
+
+            /// <summary>
+            /// A list of all substances in this container
+            /// </summary>
+            private List<SubstanceEntry> _substances;
+
+            public ReadOnlyCollection<SubstanceEntry> Substances => _substances.AsReadOnly();
+
+            private SubstanceContainerActor _actor;
+
+            /// <summary>
+            /// Is the container locked?
+            /// </summary>
+            private bool _locked;
+
+            /// <summary>
+            /// The temperature of the container
+            /// </summary>
+            private float _temperature;
+
+            /// <summary>
+            /// The total number of moles
+            /// </summary>
+            public float TotalMoles => Substances.Sum(x => x.Moles);
+
+            /// <summary>
+            /// The remaining volume in milliliters that fit in this container
+            /// </summary>
+            public float RemainingVolume => _volume - _currentVolume;
+
+            public float Temperature => _temperature;
+
+
+            /// <summary>
+            /// The capacity of this container in milliliters
+            /// </summary>
+            public float Volume => _volume;
+
+            public bool IsEmpty => _currentVolume == 0f;
+
+            public bool CanTransfer => !_locked;
+
+            /// <summary>
+            /// Multiplier to convert moles in this container to volume
+            /// </summary>
+            public float MolesToVolume
+            {
+                get
+                {
+                    float val = 0;
+                    float total = TotalMoles;
+                    foreach (var entry in Substances)
+                    {
+                        val += entry.Substance.MillilitersPerMole * (entry.Moles / total);
+                    }
+                    return val;
```

Comment:
I'd move this to a Getter function, its frowned upon to make big getters with logic in them.